torch/_functorch/vmap.py

```diff
@@ -125,7 +125,7 @@ def _process_batched_inputs(
 def _create_batched_inputs(
         flat_in_dims: List[Any], flat_args: List[Any], vmap_level: int, args_spec) -> Tuple:
     # See NOTE [Ignored _remove_batch_dim, _add_batch_dim]
-    batched_inputs = [arg if in_dim is None else
+    batched_inputs = [arg if (in_dim is None or arg is None) else
```

Comment:
This might be wrong, maybe we should handle this at the autograd Function level

Reply:
@albanD @zou3519 
I'm not sure how hard it is to add support for this at the functorch compatible autograd Function level. From looking at the ops in autograd_function_db it looks like we currently assume all grad_outputs to be non-none? The only op in autograd_function_db that takes in a None for its backward is xfailed everywhere (NumpyCubeNotComposableAutogradFunction), and we also don't support set_materialize_grads either.

I can think more about it if needed, but if it is non-trivial to add support for functorch, I can imagine it being difficult for certain other users as well. How do we feel about just adding a new config to opt-out restore back to materialization of grad_outputs?

Reply:
We shouldn't change the vmap behavior. We will need to think more about what the vmap behavior on None is (it involves pytrees).

A simple fix is to modify `restore_vmap` to handle this situation. The implication is also that `generate_vmap_rule=True` doesn't handle `ctx.set_materialize_grads(False)` correctly (looking through autograd_function_db, it looks like we didn't handle it).

Though we might want to be careful around changing the default:
- Many users will assume that the grad_outputs are always Tensors. This would break them.
- One potential way around this is to add a new ctx config. I'd be fine with that.
- Another option is to add this behavior to ctx.set_materialize_grads ? I would have expected ctx.set_materialize_grads to already control this behavior, but it sounds like it doesn't

Reply:
could it instead make these grads into meta tensors or sth like that? so that they are not None, but not materialized? or EfficientZeroTensors? or just a scalar zero broadcasted to the target shape? these should not take much space